lib/email/message_id_service.rb

```diff
@@ -58,13 +62,54 @@ def generate_for_topic(topic, use_incoming_email_if_present: false, canonical: f
         end
       end
 
+      ##
+      # The outbound_message_id may be present because either:
+      #
+      # * The post was created via incoming email and Email::Receiver, and
+      #   references a Message-ID generated by an external email client or service.
+      # * At least one email has been sent because of the post being created
+      #   to inform interested parties via email.
+      #
+      # If it is blank then we should assume Discourse was the originator
+      # of the post, and generate a Message-ID to be used from now on using
+      # our discourse/post/POST_ID@HOST format.
+      def generate_or_use_existing(post)
+        if post.outbound_message_id.blank?
+          post.update(outbound_message_id: "discourse/post/#{post.id}@#{host}")
+        end
+        "<#{post.outbound_message_id}>"
+      end
+
+      def generate_or_use_existing_bulk(post_ids)
+        return [] if post_ids.empty?
+
+        DB.query_single(<<~SQL, host: host)
+          UPDATE posts
+          SET outbound_message_id = 'discourse/post/' || posts.id || '@' || :host
+          WHERE outbound_message_id IS NULL AND posts.id IN (#{post_ids.join(",")});
+          SELECT '<' || posts.outbound_message_id || '>'
+          FROM posts
+          WHERE posts.id IN (#{post_ids.join(",")})
+          ORDER BY posts.created_at ASC;
+        SQL
```

Comment:
One first look, I didn't quite understand what this query was doing until I realised there are two queries being executed here. I think it'll be easier to read if the queries are split up into `DB.exec` for the update statement and then a `DB.query_single` for the select statement.

Reply:
That's fair enough, I just thought I would save the round trip by just doing one, but if it's easier to read I'll just do a second query :+1:

Reply:
I realised that we do not just improve readability here but also ensures that if one of the query fails, it'll be easier to pin point which query. Right now I'm not sure how the error will look if the update statement fails. I'm guessing we'll get a PG error with the relevant message.

Reply:
Yep all fair enough, will change.